lib/firenest/synced_server.ex

```diff
@@ -0,0 +1,453 @@
+      end
+    else
+      {:ok, state}
+    end
+  end
+
+  @impl true
+  def format_status(:normal, [pdict, %{mod: mod, int: int}]) do
+    try do
+      apply(mod, :format_status, [:normal, [pdict, int]])
+    catch
+      _, _ ->
+        [{:data, [{'State', int}]}]
+    else
+      mod_status ->
+        mod_status
+    end
+  end
+
+  def format_status(:terminate, [pdict, %{mod: mod, int: int}]) do
+    try do
+      apply(mod, :format_status, [:terminate, [pdict, int]])
+    catch
+      _, _ ->
+        int
+    else
+      mod_state ->
+        mod_state
+    end
+  end
+
+  defp handle_common(result, state) do
+    case result do
+      {:noreply, int} -> {:noreply, %{state | int: int}}
+      {:noreply, int, timeout} -> {:noreply, %{state | int: int}, timeout}
+      {:stop, reason, int} -> {:stop, reason, %{state | int: int}}
+      other -> {:stop, {:bad_return_value, other}, state}
+    end
+  end
+
+  defp init_mod(mod, arg, state) do
+    Process.put(:"$initial_call", {mod, :init, 2})
+
+    try do
+      mod.init(arg)
+    catch
+      :throw, value ->
+        reason = {{:nocatch, value}, System.stacktrace()}
+        {:stop, reason}
+    else
+      {:ok, int} -> {:ok, %{state | int: int}}
+      {:ok, int, timeout} -> {:ok, %{state | int: int}, timeout}
+      {:stop, reason} -> {:stop, reason}
+      :ignore -> :ignore
+      other -> {:stop, {:bad_return_value, other}}
+    end
+  end
+
+  defp sync_named(topology, name, state) do
+    node_ref = Topology.node(topology)
+
+    case Topology.sync_named(topology, self()) do
+      {:ok, replicas} ->
+        init_replicas(topology, name, replicas, node_ref)
+        {:ok, %{state | awaiting_hello: replicas}, node_ref}
+
+      {:error, error} ->
+        {:stop, {:sync_named, error}}
+    end
+  end
+
+  defp init_replicas(topology, name, replicas, node_ref) do
+    Enum.each(replicas, &Topology.send(topology, &1, name, {__MODULE__, :hello, node_ref}))
+  end
+
+  defp apply_callback(%{mod: mod, int: int}, fun, args) do
+    try do
+      apply(mod, fun, args ++ [int])
+    catch
+      :throw, value ->
+        :erlang.raise(:error, {:nocatch, value}, System.stacktrace())
+
+      :error, :undef when fun == :handle_info ->
+        if function_exported?(mod, :handle_info, 2) do
+          :erlang.raise(:error, :undef, System.stacktrace())
+        else
+          proc =
+            case Process.info(self(), :registered_name) do
+              {_, []} -> self()
+              {_, name} -> name
+            end
+
+          pattern = "Undefined handle_info/2 in ~p, process ~p received unexpected message: ~p~n"
+          :error_logger.warning_msg(pattern, [mod, proc, hd(args)])
+          {:noreply, int}
+        end
+
+      :error, :undef when fun == :handle_replica ->
+        if function_exported?(mod, :handle_replica, 3) do
+          :erlang.raise(:error, :undef, System.stacktrace())
+        else
+          {:noreply, int}
+        end
```

Comment:
I'm not a huge fan of doing this like that, but that's more-or-less how it's done in `gen_server`. Any ideas how to improve?